File: common/java/src/com/db/common/ConfigFile.java
/*
 * Copyright (c) 2005-2006 Digital Bazaar, Inc.  All rights reserved.
 */
package com.db.common;

import com.db.logging.Logger;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.Iterator;
import java.util.Properties;

/**
 * A class for managing a configuration file.
 * 
 * @author Dave Longley
 */
public class ConfigFile extends ConfigOptions
{
   /**
    * The name of the config file.
    */
   protected String mFilename;
   
   /**
    * A header to write in the file.
    */
   protected String mHeader;
   
   /**
    * Creates a new config file.
    * 
    * @param filename the name of the config file.
    */
   public ConfigFile(String filename)
   {
      mFilename = filename;
      mHeader = null;
   }
   
   /**
    * Loads the configuration settings from a file.
    *
    * @return true if successful, false if not.
    */
   public synchronized boolean read()
   {
      boolean rval = false;
      
      mProperties = new Properties();
   
      try
      {
         // read the configuration from a file
         mProperties.load(new FileInputStream(mFilename));
         rval = true;
      }
      catch(Throwable t)
      {
         getLogger().error(
               "Could not read config file: \"" + mFilename + "\"");
         getLogger().debug(Logger.getStackTrace(t));
      }

      return rval;
   }
   
   /**
    * Writes the configuration settings to a file.
    *
    * @return true if successful, false if not.
    */
   public boolean write()
   {
      return write(mHeader);
   }

   /**
    * Writes the configuration settings to a file.
    *
    * @param header a header to include at the top of the file.
    * @return true if successful, false if not.
    */
   public synchronized boolean write(String header)
   {
      boolean rval = false;
      
      try
      {
         // write the configuration to a file
         mProperties.store(new FileOutputStream(mFilename), header);
         rval = true;
      }
      catch(Throwable t)
      {
         getLogger().error(
               "Could not write config file: \"" + mFilename + "\"");
         getLogger().debug(Logger.getStackTrace(t));
      }

      return rval;
   }
   
   /**
    * Sets the header to write to the file, if any.
    * 
    * @param header the header to write to the file. 
    */
   public void setHeader(String header)
   {
      mHeader = header;
   }
   
   /**
    * Writes the configuration in the passed config options to this
    * config file.
    * 
    * @param config the configuration options to write to disk.
    * @return true if successfully written, false if not.
    */
   public boolean writeConfig(ConfigOptions config)
   {
      boolean rval = false;
      
      Iterator i = config.getKeys().iterator();
      while(i.hasNext())
      {
         String key = (String)i.next();
         setValue(key, config.getString(key));
      }
      
      rval = write();
      
      return rval;
   }
   
   /**
    * Reads from this configuration file and populates the passed
    * configuration options with the read options. 
    * 
    * @param config the configuration options to populate.
    * @param read true if the configuration should be read from file,
    *        false if it should not be.
    */
   public void populateConfig(ConfigOptions config, boolean read)
   {
      // read from file
      if(read)
      {
         read = read();
      }
      else
      {
         read = true;
      }
      
      if(read)
      {
         // copy all values read from file into the configuration
         Iterator i = getKeys().iterator();
         while(i.hasNext())
         {
            String key = (String)i.next();
            config.setValue(key, getString(key));
         }
      }
   }
}

File: obsolete/common/csharp/DB.Common/ConfigFile.cs
/*
 * Copyright (c) 2005 Digital Bazaar, Inc.  All rights reserved.
 */
using System;
using System.Collections;
using System.Globalization;
using System.IO;
using System.Runtime.CompilerServices;

using DB.Common.Logging;

namespace DB.Common
{
   /// <summary>
   /// A class for managing a configuration file.
   /// </summary>
   ///
   /// <author>Dave Longley</author>
   public class ConfigFile : ConfigOptions
   {
      /// <summary>
      /// The name of the config file.
      /// </summary>
      protected string mFilename;
      
      /// <summary>
      /// A header to write in the file.
      /// </summary>
      protected string mHeader;
      
      /// <summary>
      /// Creates a new config file.
      /// </summary>
      /// 
      /// <param name="filename">the name of the config file.</param>
      public ConfigFile(string filename)
      {
         mFilename = filename;
         mHeader = null;
      }
      
      /// <summary>
      /// Loads the configuration settings from a file.
      /// </summary>
      ///
      /// <returns>true if successful, false if not.</returns>
      [MethodImpl(MethodImplOptions.Synchronized)]
      public virtual bool Read()
      {
         bool rval = false;
         
         mOptions = new Hashtable();
         
         try
         {
            // read the configuration from a file
            TextReader tr = File.OpenText(mFilename);
            
            string line = null;
            while((line = tr.ReadLine()) != null)
            {
               // only parse the line if it starts with a comment character
               if(!line.StartsWith("#"))
               {
                  // read to the first equals
                  int equals = line.IndexOf("=");
                  if(equals != -1)
                  {
                     string key = line.Substring(0, equals);
                     string val = "";
                     if((equals + 1) < line.Length)
                     {
                        // set the value to the rest of the line
                        val = line.Substring(equals + 1,
                                             line.Length - (equals + 1));
                     }
                     
                     mOptions[key] = val;
                  }
                  else
                  {
                     throw new System.FormatException(
                     "Configuration option is not of the format \"key=value\"");
                  }
               }
            }
            
            rval = true;
         }
         catch(Exception e)
         {
            Logger.Error("Could not read config file: \"" + mFilename + "\"");
            Logger.Debug(e.ToString());
         }
         
         return rval;
      }
      
      /// <summary>
      /// Writes the configuration settings to a file.
      /// </summary>
      /// 
      /// <returns>true if successful, false if not.</returns>
      public virtual bool Write()
      {
         return Write(mHeader);
      }
      
      /// <summary>
      /// Writes the configuration settings to a file.
      /// </summary>
      /// 
      /// <param name="header">header a header to include at the top of
      /// the file.</param>
      /// <returns>true if successful, false if not.</returns>
      [MethodImpl(MethodImplOptions.Synchronized)]
      public virtual bool Write(string header)
      {
         bool rval = false;
         
         try
         {
            // delete existing config file
            File.Delete(mFilename);

            // write the configuration to a file
            TextWriter tw = File.CreateText(mFilename);
            
            // write header to file if one exists
            if(Header != null)
            {
               tw.WriteLine("# " + mHeader);
            }
            
            // write date to file
            string df = "M/d/yyyy h:mm:ss tt 'GMT' z";
            DateTime now = DateTime.Now;
            string date = now.ToString(df, DateTimeFormatInfo.InvariantInfo);
            tw.WriteLine("# " + date);
            
            // write the key-value pairs to file
            foreach(string key in Keys)
            {
               tw.WriteLine(key + "=" + GetString(key));
            }
            
            rval = true;
         }
         catch(Exception e)
         {
            Logger.Error("Could not write config file: \"" + mFilename + "\"");
            Logger.Debug(e.ToString());
         }
         
         return rval;
      }
      
      /// <summary>
      /// Gets/Sets the header to write to the file, if any.
      /// </summary>
      public virtual string Header
      {
         get
         {
            return mHeader;
         }
         
         set
         {
            mHeader = value;
         }
      }
   }
}
